README.md
# devproxy - HTTP(S) proxy for developers

[<!--lint ignore no-dead-urls-->![GitHub Actions status | rewgaz/devproxy setup](https://github.com/rewgaz/devproxy/workflows/setup/badge.svg)](https://github.com/rewgaz/devproxy/actions?workflow=setup)

HTTP(S) proxy config builder - to make development with multiple containers easier.

## Preface

Working with multiple Docker containers on different projects running on port 80/443 at the same time requires a lot of annoying configuration and potentially ends in a headache. This project tries to make this process a lot easier and faster, and makes it look like you only have one webserver running.

`!! Use this tool at your own risk. If your system setup differs from common defaults, this tool might not work or even break your system's configuration. !!`

Supported operating systems:

- Fedora (Tested on Fedora Server 30)
- Ubuntu (Tested on Ubuntu 18.04)

Note: Most of the commands need root privileges.

## Setup - Step 1: Initial setup

Note: Most of the commands need root privileges.

Install devproxy:

```bash
make install
```

The setup command disables a potentially running `Apache Webserver` and installs `NGINX` and `certbot`.
It also opens port `80` and `443` in your firewall if necessary.

```bash
devproxy setup
```

## Setup - Step 2: Host configuration

Put a config file for each host in the `/etc/devproxy/config` directory and name it for example `my-website.com.json`. Only JSON files with the suffix `.json` will be read.

### Example

This example config file is for a webserver using HTTP and HTTPS. The host names `my-website.com` and `alternative.my-website.com` will be redirected to `www.my-website.com`. All incoming traffic on `www.my-website.com` will be passed to a webserver listening on `0.0.0.0:18010` for HTTP and `0.0.0.0:14010` for HTTPS.

```json
{
    "version": 1,
    "host": {
        "hostname": "www.my-website.com",
        "alternatives": [
            "my-website.com"
            "alternative.my-website.com"
        ],
        "targets": [
            {
                "protocol": "http",
                "ip": "0.0.0.0",
                "port": "18010"
            },
            {
                "protocol": "https",
                "ip": "0.0.0.0",
                "port": "14010",
                "ssl_cert": "/etc/letsencrypt/live/my-website.com/fullchain.pem",
                "ssl_key": "/etc/letsencrypt/live/my-website.com/privkey.pem"
            }
        ]
    }
}
```

This host config will result in the following behaviour:

```plain
http://my-website.com -> (301) -> http://www.my-website.com
http://alternative.my-website.com -> (301) -> http://www.my-website.com
http://www.my-website.com -> (proxy_pass) -> http://0.0.0.0:18010

https://my-website.com -> (301) -> https://www.my-website.com
https://alternative.my-website.com -> (301) -> https://www.my-website.com
https://www.my-website.com -> (proxy_pass) -> https://0.0.0.0:14010
```

## Setup - Step 3: Apply config

The `apply` command parses all config files and creates all necessary NGINX config files and adds the hostnames to your `/etc/hosts` file. (A copy of the old `/etc/hosts` file will be saved to `/etc/devproxy/hosts-backup`.) Rerun this command after you changed any config files or added/renewed any SSL certificates.

```bash
devproxy apply
```

## SSL certificates

To quickly generate SSL certificates for your hosts you can use the buildin Certbot commands.

### Create a new SSL certificate

Create a SSL certificate with Certbot for your host.

```bash
devproxy cert:create my-website.com
```

### Renew SSL certificates

Renew all SSL certificates created with Certbot if necessary.

```bash
devproxy cert:renew
```

### Fedora: Use Certbot certificate for Cockpit

Configure Cockpit to use a certificate created by the `cert:create` command. This will override the current SSL certificate in Cockpit if present.

```bash
devproxy cert:cockpit.sh my-website.com
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.1 - Unreleased]

## [0.1.0]

### Added

- Makefile for installation and uninstallation
- HTML error templates for NGINX
- Setup command that disables Apache webserver and installs NGINX and Certbot
- NGINX config builder
- Command to apply the configuration to the system
- Certbot helper command to create and renew Let's Encrypt SSL certificates
- Command to use Certbot certificate for Cockpit on Fedora
- Help command
- Github action setup workflow
- Readme
- Changelog
- MIT license

[0.1.1 - Unreleased]: https://github.com/rewgaz/devproxy
[0.1.0]: https://github.com/rewgaz/devproxy/releases/tag/0.1.0


src/builder.py
# -*- coding: utf-8 -*-
import glob
import json
import os
import shutil


path_config = '/usr/src/http-gateway-config'
path_build = '/usr/src/http-gateway-build'
path_hosts = '/usr/src/http-gateway-build/hosts'
hosts_separator = '### http-gateway ###'

template_location = '''location / {{
        proxy_pass {protocol}://{ip}:{port}/;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    }}'''

template_http = '''server {{
    listen 80;
    listen [::]:80;
    server_name {server_name};

    '''+template_location+'''
    
    {redirects}
}}'''

template_https = '''server {{
    listen 443 ssl;
    listen [::]:443 ssl;
    server_name {server_name};

    ssl_certificate "{ssl_cert}";
    ssl_certificate_key "{ssl_key}";
    ssl_session_cache shared:SSL:1m;
    ssl_session_timeout  10m;
    ssl_prefer_server_ciphers on;
    ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers HIGH:!aNULL:!MD5;
    keepalive_timeout 70;

    '''+template_location+'''
    
    {redirects}
}}'''

template_redirect_to_https = '''if ($scheme != "https") {
        return 301 https://$host$request_uri;
    }'''

template_redirect_to_host = '''if ($host != "{host}") {{
        return 301 {protocol}://{host}$request_uri;
    }}'''


def read_config_ver_1_0(config_file):
    """
        Read host config file and return data
        :param config_file: string
        :return: dict
    """
    parser_version = '1'

    data = json.load(config_file)
    file_version = 'unknown'
    if isinstance(data, dict) and data.get('version') is not None:
        file_version = str(data.get('version'))

    if file_version != parser_version:
        raise Exception('Config file version mismatch: Expected \'{parser_version}\', got \'{file_version}\''.format(
            parser_version=parser_version,
            file_version=file_version,
        ))

    return data


def read_config():
    """
    Read all host config files and return as list
    :return: list
    """
    config = []
    config_files = glob.glob(path_config+'/*.json')

    for config_filename in config_files:
        config_file = open(config_filename, 'rt')
        config.append(read_config_ver_1_0(config_file))
        config_file.close()
    
    return config


def build_hosts_str(host):

    # todo alternatives
    return '127.0.0.1 '+host['host']['hostname'].strip()+'\n'


def build_nginx_conf(host):
    """
    Write NGINX config for a host
    """
    hostname = host['host']['hostname'].strip()
    hostnames_all = hostname
    for alternative in host['host']['alternatives']:
        hostnames_all = hostnames_all + ' ' + alternative.strip()

    for target in host['host']['targets']:

        redirects = ''
        if len(host['host']['alternatives']) > 0:
            redirects = template_redirect_to_host.format(
                host=hostname,
                protocol=target['protocol'],
            )

        if target['protocol'] == 'http':
            host_config = template_http.format(
                server_name=hostnames_all,
                protocol='http',
                ip=target['ip'].strip(),
                port=target['port'].strip(),
                redirects=redirects,
            )

        elif target['protocol'] == 'https':
            host_config = template_https.format(
                server_name=hostnames_all,
                protocol='https',
                ip=target['ip'].strip(),
                port=target['port'].strip(),
                ssl_cert=target['ssl_cert'].strip(),
                ssl_key=target['ssl_key'].strip(),
                redirects=redirects,
            )

        else:
            raise Exception('Unrecognized protocol \'{protocol}\' @ {hostname}'.format(
                hostname=hostname,
                protocol=target['protocol'],
            ))

        config_file = open(path_build + '/conf.d/' + hostname + '.' + target['protocol'] + '.conf', 'xt')
        config_file.write(host_config)
        config_file.close()


def build_ssl_cert(host):

    path_cert = path_build + '/ssl/' + host['hostname'] + '.pem'
    path_key = path_build + '/ssl/' + host['hostname'] + '.key'

    if host['protocol'] == 'https' and not os.path.isfile(path_cert):
        cmd = '''
            openssl req -x509 -nodes -days 365 -newkey rsa:2048 \
            -out '''+path_cert+''' \
            -keyout '''+path_key+''' \
            -subj "/C=AT/ST=Vienna/L=Vienna/O=Security/OU=Development/CN=localhost" \
        '''
        os.system(cmd)


def build_hosts_file(hosts_data, hosts_file_lines):
    """
    Write new /etc/hosts file
    """

    first_line_to_ignore = None
    for num, line in enumerate(hosts_file_lines, start=0):
        if line.startswith(hosts_separator):
            first_line_to_ignore = num
            break

    if first_line_to_ignore is None:
        hosts_data_all = hosts_file_lines
    else:
        hosts_data_all = hosts_file_lines[0:first_line_to_ignore]

    hosts_data_all.append('\n' + hosts_separator + '\n')
    hosts_data_all = hosts_data_all + hosts_data

    hosts_file = open(path_hosts, 'wt')
    for line in hosts_data_all:
        hosts_file.write(line)
    hosts_file.close()


def build():
    """
    Entry point
    """

    # Read existing /etc/hosts file
    hosts_file = open(path_hosts, 'rt')
    hosts_file_lines = hosts_file.readlines()
    hosts_file.close()

    # Remove all files in build folder
    shutil.rmtree(path_build, ignore_errors=True)
    os.makedirs(path_build + '/conf.d/', exist_ok=True)
    os.makedirs(path_build + '/ssl/', exist_ok=True)

    # Read all host config files
    config = read_config()
    hosts_data = []
    for host in config:
        build_nginx_conf(host)
        #build_ssl_cert(host)
        hosts_data.append(build_hosts_str(host))

    # Write new /etc/hosts file
    build_hosts_file(hosts_data, hosts_file_lines)


if __name__ == '__main__':
    build()
